### ISSUE TITLE
DSA signature verification issue

### ISSUE THREAD

**Issue by jbarnesweb** [usnistgov/ACVP-Server#113]
**environment**
Demo

**testSessionId**
Unknown

**vsId**
563159

**Algorithm registration**

```
			{
				"algorithm": "DSA",
				"mode": "sigVer",
				"revision": "1.0",
				"prereqVals": [
					{
						"algorithm": "SHA",
						"valValue": "same"
					},
					{
						"algorithm": "DRBG",
						"valValue": "same"
					}
				],
				"capabilities": [
					{
						"n": 256,
						"l": 2048,
						"hashAlg": [
							"SHA2-256"
						]
					},
					{
						"n": 256,
						"l": 3072,
						"hashAlg": [
							"SHA2-256"
						]
					}
				]
			},

```

**Endpoint in which the error is experienced**
post vector set results (unknown test session id)

**Expected behavior**
All test cases pass.

**Actual behavior**
One test case fails

**Additional context**
The test case causing the failure.

```
      {
        "g": "382635D655B4E8F59181F9174BBFCE8839257C4296B73C05B0E088354275B084B303C37E6CEDDD7F158F986CF60D6E4EF1EECB86E01465F7580D5D1A106D616D6EBE43C84184062424A6F2B84ECD5029DE3FC8F9A29F5350B391A103524CB04F95D1FB1680F4802E7CA03853436D6E4DEAFCF3ACE616FC173CBAC9539A4795D4BA70BBDE3F5116B76CAB8463BFEF81DAEA0007971E7844B203FBAE2554362DF530000F4B73FDFCE5E2D4FC9DB2B9B6BFE441CC278EF5E82153DE3F3A3BAE6C43B2732F8FDDC6E89B30AF6AEBB3751982B7CB60DC037FD5641850420DDFEF54CF720237428C10B6DAB51C2843CB6788DEEA7ABCBA2CD3BA4367B8513BC90871A07BAD504D899122B6BC37475E3854FCC0C05B6E761354DCBFA3406728CBF3C021ED223A1688AFBFBBAE0DCFD6AFDC224D9998C039161AC4D3C39E8DD5457C97E0C23310CC3A5CF9C781E595C55A8919196417AE306521A2E1981B9CC52BF302ADC650F2C9FE6E4E9A507CA5630417F4A200B1EB59C1B4A204E76DBA455C99956D",
        "hashAlg": "SHA2-256",
        "l": 3072,
        "n": 256,
        "p": "A7E1D2B2976420B86961C071A917AC70220BF3E3723370E081D40F54545AD5CF4262799357962CC3446767B2735CC0656852F8BDB116A947062E43B970F5E6B52B3BF69057775121CD153F14A9380C01866718F6786EA06545A8DB88565B04324104B26208EB0FFF8FAD46333DF88636B04A5387FEF723F59D7E357E3EC8C843C507A5D7DB52DDEB08B9AE0DC585EC2DA3574FC74DEF3DD83D239C6A04691C8DCD4916E88D96D04E4C4EEF73BBB40171A3F8EB1F021D0B6FEDEC643059916B0024F455FECB1C84016428C466130D9BCCFAA33945BFD80D4A27818C4C8DC56D8BD1CAD3278688F142BFCA17725B76C4A31FA79A1773EC9D79BCCD8E377914AB8C54980B08063B21E8A1D5899102DFDABB633A97D2162B072D204B4469FCDC03D9FB8BB2BCDB42B45ABD035ECCFB8571BD6107E50895197CEFECE5EB66F9714F185497F8BBFF68C52AFDE6670AC8609767DB0D55CBDD798C47CC6836B4972B25CFAEF22C4DD36010DDD49544C0EE31C636AF66C722CF1C0F940597322D6561474F",
        "q": "A02D08803FB3C9FAEFFD75B9A21D07AE0F92D68AAC01E87D0DDE9487427AACBF",
        "testType": "AFT",
        "tests": [
...
          {
            "message": "5FE6571A4801FB2FC3C052CED48BB3BE92BE426D82E35D5C7FDD703AD09873DA",
            "r": "49E7B6EDD9C4ADB8C4DEB9DEF3B546E2ECD25DA3C462F97464A4F1A40CAF0468",
            "s": "CDE8B0D32B81942FA057642C811345407BCCB425E03FF44A9BFCBBE06D7D6A",
            "tcId": 28,
            "y": "A6BC39964E7C02C8E51496CBA6D42096E4CD216043B3179B8CE9C821087CA0873D8BB933D2102FA43016901482B7F9E5B9F30FD21B12D28C6A185516882313C7F83B10D6DC53B391A4207E806FEA9CD2B9577C27C0587CE46E1A61DA7CAC7C28A5EE65F36BEEBAC66720770FF72AD31ABF324EFE0D51F7F55766ACBEA0296AA45DD5B6539A54771B9F6834A8D8CDE1BD0A88634CA4077768B47B2C2B98C39CCAD0F1B9F37A5C8994C1B2129C9C2E92FEA64F9BED5D196905DD444ADB28D971668525039705A8E42DA229E2D6234258F6F63BC815FCB976685E54B17978F3B14FDBA96E86EA135BAA72195DB643CB23F31FE56E0830B7C61D62945CBA20E5615D17F048E23C54F601E4122A81EAD075C6CB962313A84750B2C8D2098BC9BB1FB8CD6232F422A7A932CB8DB7F758B978323ECF0091845C5A2C1C0B9F90CA7E45A392FF6F742B941F1550531539A4266A63F59C283BEE1C4934BD86AB27BFF17A20199EA9A2B1FCF4BC653AE78E9CBEFA12486F7BCBB613184FA385C203B8B0B2D1"
          },
...
        }
...

```

We mark the verification as a failure.

I ran the same test data through our legacy CAVS tool and got the same result (verification failure).

I recreated the public key from p, q, g and y values provided using OpenSSL. I recreated a signature file with the given r and s values. I ran the following OpenSSL command.

```
$ openssl dgst -verify dsapub.der -keyform der -signature dsasig.der message.txt.bin
Error Verifying Data
```

That response prompted me to look at the key.

```
$ openssl dsa -in dsapub.der -pubin -inform der -noout -text
pub:  (Negative)
    59:43:c6:69:b1:83:fd:37:1a:eb:69:34:59:2b:df:
    69:1b:32:de:9f:bc:4c:e8:64:73:16:37:de:f7:83:
...

```

Negative and the pub: doesn't look anything like y.

 FIPS 186-4 says the following in Section 4.1.

"y the public key, where y = g^x mod p."

If I understand that correctly, y should never be negative. So I tried prepending a 0x00 byte to y and it passed. I can't find anywhere in FIPS 186-4 that prescribes prepending a zero byte to import a negative key.

Is this faulty data?

**Comment by jbarnesweb** [comment]
ping

**Comment by Kritner** [comment]
@jbarnesweb I ran the tcId 28 called out in the issue against our own crypto and am getting a successful verification of the signature.  To double check that it wasn't our crypto as fault, I implemented the signature verify process in python, and ran both a "passing" (from both of our perspectives) and "failing" (from your perspective) signature verification against it, and they both continue to be valid.

https://replit.com/@Kritner/VerifyDsaSignature#main.py

-----

I should have read the entire issue more closely as the second to the last sentence would have saved me a fair amount of time on trying to reproduce 😅 

An empty byte (or more) is added (or should be added) as a part of padding on the client end to cancel out the 2s complement, because as you said, y should not be interpreted as a negative number.  It's been a while since I've looked, but AFAIK we generally leave it up to the IUT to cancel out the 2s complement.

I'm on vacation now but hopefully @livebe01 or @celic can field further concerns; or confirm/deny my current interpretation.
